[ai-text-analytics] Unify TA error model

willmtemple opened the pull request:
This PR adds a convenience layer to the text analytics error model that flattens out the `innerError` hierarchy, since TA only supports one nested inner error.

Effectively, we are swallowing the generated TextAnalyticsError and InnerError and replacing them with a single TextAnalyticsError that is not generated.

We override the ErrorCode enum because there is a discrepancy between the string literals specified in the swagger and the string literals that we observe during testing. The new model has only `code`, `message`, and an optional `target` (though I couldn't find a scenario in which I could make the service return anything other than `target: undefined`.

I also fixed a couple of typos (makeTextAnalysisResult -> makeTextAnalyticsSuccessResult and makeTextAnalysisErrorResult -> makeTextAnalyticsErrorResult).

willmtemple commented:
/azp run js - textanalytics - tests

willmtemple commented:
> This looks good to me. I'm assuming that you have tests for this merged with another PR?

no I just hadn't written them yet, was still working on them @kristapratico

willmtemple commented:
/azp run js - textanalytics - tests

xirzec commented on sdk/textanalytics/ai-text-analytics/src/textAnalyticsResult.ts:
```diff
@@ -1,13 +1,70 @@
+ * An Error Code returned from the Text Analytics service. Possible
+ * values include:
+ *
+ * - 'InvalidRequest'
```
this comment feels weird since the type is a string literal union...

willmtemple replied:
My thought was that it could be useful for JS but then I realized that just because a user is using JS doesn't mean they don't see the full typed intellisense or our beautiful ref docs.

Once the docs.ms docs are rendered we can check this again because I'm not 100% sure it will show correctly there.

xirzec commented on sdk/textanalytics/ai-text-analytics/src/textAnalyticsResult.ts:
```diff
@@ -57,12 +138,12 @@ export function makeTextAnalysisResult(
   };
 }
 
-export function makeTextAnalysisErrorResult(
```
😳 oops

xirzec commented on sdk/textanalytics/ai-text-analytics/test/error.spec.ts:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Works around an issue with the swagger models and casing
+ */
+function cap<T extends ErrorCodeValue | InnerErrorCodeValue>(code: T): T {
```
I'm confused by why this function is needed - is the enum type that is generated by swagger incorrect? If so, it seems like we should fix that rather than working around it?

willmtemple replied:
This was related to the issue with the swagger but I've just patched the swagger using the README in the latest commits.

maririos commented on sdk/textanalytics/ai-text-analytics/review/ai-text-analytics.api.md:
```diff
@@ -86,7 +86,7 @@ export interface Entity {
 
 // @public
-export type ErrorCodeValue = 'invalidRequest' | 'invalidArgument' | 'internalServerError' | 'serviceUnavailable';
+export type ErrorCode = "InvalidRequest" | "InvalidArgument" | "InternalServerError" | "ServiceUnavailable" | "InvalidParameterValue" | "InvalidRequestBodyFormat" | "EmptyRequest" | "MissingInputRecords" | "InvalidDocument" | "ModelVersionIncorrect" | "InvalidDocumentBatch" | "UnsupportedLanguageCode" | "InvalidCountryHint";
```
idk much about JS, but what happens if the service adds a new `ErrorCode` ?

willmtemple replied:
I addressed this in the latest commit. This line was only meant to be a workaround for the swagger bug but I have patched it in the generated code.

maririos commented on sdk/textanalytics/ai-text-analytics/src/textAnalyticsResult.ts:
```diff
@@ -1,13 +1,70 @@
 /**
  * The result of a text analytics operation on a single input document.
  */
 export type TextAnalyticsResult = TextAnalyticsSuccessResult | TextAnalyticsErrorResult;
```
In the other languages we have a helper property `isError` or `HasError` . Will a helper like this make sense in JS? 
that way a user can do `if (!errorResult.isError)` instead of `if (errorResult.error === undefined)`

willmtemple replied:
In JS we can just do

```js
if (result.error) {
  // result is sure to be a success variant here
}
```

since something that is `undefined` is falsy.

For users using TypeScript, they will even get a strong type bound for `result: TextAnalyticsSuccessResult` in the body of the if, which is why we have the `error?: undefined` in the definition of `TextAnalyticsSuccessResult`. It is "defined to be undefined" for the success type.

maririos replied:
Cool. Thanks for explaining

willmtemple commented:
> Really a bit confused about what's going on with the error codes, I know you mentioned it to me offline, but I'd like a refresher. sweat_smile

(@xirzec )

The issue is that the swagger specifies the values of the `ErrorCodeValue` and `InnerErrorCodeValue` to be in camelCase, while they are actually produced by the service in PascalCase. TS is the only of the four languages that has the capability to model this value in the type system more strictly than as a mere string, and so we're the only ones who hit this.

In the latest commit, I just applied the capitalization as a swagger transform. If the service swagger is fixed, we can just remove it.

I also removed the manual code enumeration and replaced it with 

```typescript
export type ErrorCode = ErrorCodeValue | InnerErrorCodeValue;
```

willmtemple commented:
/azp run js - textanalytics - tests

willmtemple commented:
@xirzec I'm merging this now after talking it over with Ramya and so I can send a build to the service team with the new API. If you take a look and see anything that jumps out at you, I can issue a quick patch on Monday

xirzec commented:
@willmtemple the final version of this PR looks great! Thanks for the changes. 👍
